File: lib/py/uapi/internal/binary/PackList.py
from typing import TYPE_CHECKING

from msgpack import ExtType

if TYPE_CHECKING:
    from uapi.internal.binary.BinaryPackNode import BinaryPackNode


PACKED_BYTE = 17


def pack_list(lst: list[object]) -> list[object]:
    from uapi.internal.binary.Pack import pack
    from uapi.internal.binary.PackMap import pack_map
    from uapi.internal.binary.CannotPack import CannotPack

    if not lst:
        return lst

    packed_list: list[object] = []
    header: list[object] = []

    packed_list.append(ExtType(PACKED_BYTE, b''))

    header.append(None)

    packed_list.append(header)

    key_index_map: dict[int, BinaryPackNode] = {}
    try:
        for item in lst:
            if isinstance(item, dict):
                row = pack_map(item, header, key_index_map)
                packed_list.append(row)
            else:
                # This list cannot be packed, abort
                raise CannotPack()
        return packed_list
    except CannotPack:
        new_list = [pack(item) for item in lst]
        return new_list

File: lib/ts/uapi/internal/binary/PackList.ts
import { BinaryPackNode } from './BinaryPackNode';
import { pack } from './Pack';
import { packMap } from './PackMap';
import { CannotPack } from './CannotPack';
import { addExtension } from 'msgpackr';

export const PACKED_BYTE = 17;

class MsgpackPacked {
    toString() {
        return 'PACKED';
    }
}
const MSGPACK_PACKED_EXT = {
    Class: MsgpackPacked,
    type: PACKED_BYTE,
    pack(instance: MsgpackPacked) {
        return Buffer.from([]);
    },
    unpack(buffer: Buffer) {
        return new MsgpackPacked();
    },
};
addExtension(MSGPACK_PACKED_EXT);

export function packList(list: any[]): any[] {
    if (list.length === 0) {
        return list;
    }

    const packedList: any[] = [];
    const header: any[] = [];

    packedList.push(new MsgpackPacked());

    header.push(null);

    packedList.push(header);

    const keyIndexMap: Map<number, BinaryPackNode> = new Map();
    try {
        for (const e of list) {
            if (e instanceof Map) {
                const row = packMap(e, header, keyIndexMap);

                packedList.push(row);
            } else {
                // This list cannot be packed, abort
                throw new CannotPack();
            }
        }
        return packedList;
    } catch (ex) {
        const newList: any[] = [];
        for (const e of list) {
            newList.push(pack(e));
        }
        return newList;
    }
}
